aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandler.java

```diff
@@ -17,42 +21,90 @@
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final Logger logger) {
 
-        final ResourceModel model = request.getDesiredResourceState();
         final String awsAccountId = request.getAwsAccountId();
-        final String idempotencyToken = request.getClientRequestToken();
+        final ResourceModel model = request.getDesiredResourceState();
+        final String pgName = model.getProfilingGroupName();
 
+        CreateProfilingGroupRequest createProfilingGroupRequest = CreateProfilingGroupRequest.builder()
+                .profilingGroupName(pgName)
+                .clientToken(request.getClientRequestToken())
+                .build();
         try {
-            CreateProfilingGroupRequest createProfilingGroupRequest = CreateProfilingGroupRequest.builder()
-                    .profilingGroupName(model.getProfilingGroupName())
-                    .clientToken(idempotencyToken)
-                    .build();
-
             proxy.injectCredentialsAndInvokeV2(createProfilingGroupRequest, profilerClient::createProfilingGroup);
+        } catch (CodeGuruProfilerException e) {
+            wrapException(e);
+        }
+        logger.log(format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, pgName, awsAccountId));
+
+        Optional<List<String>> principals = principalsForAgentPermissionsFrom(model);
+        if (principals.isPresent()) {
+            putAgentPermissions(proxy, pgName, principals.get());
+            logger.log(format("%s [%s] for accountId [%s] has been successfully updated with agent permissions!",
+                    ResourceModel.TYPE_NAME, pgName, awsAccountId));
+        }
+
+        return ProgressEvent.defaultSuccessHandler(model);
+    }
 
-            logger.log(String.format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, model.getProfilingGroupName(), awsAccountId));
+    private void putAgentPermissions(final AmazonWebServicesClientProxy proxy, final String pgName, final List<String> principals) {
+        PutPermissionRequest putPermissionRequest = PutPermissionRequest.builder()
+                .profilingGroupName(pgName)
+                .actionGroup(ActionGroup.AGENT_PERMISSIONS)
+                .principals(principals)
+                .build();
+        try {
+            proxy.injectCredentialsAndInvokeV2(putPermissionRequest, profilerClient::putPermission);
+        } catch (CodeGuruProfilerException e) {
+            try {
+                DeleteProfilingGroupRequest deletePgRequest = DeleteProfilingGroupRequest.builder().profilingGroupName(pgName).build();
+                proxy.injectCredentialsAndInvokeV2(deletePgRequest, profilerClient::deleteProfilingGroup);
+            } catch (Throwable deleteException) {
+                e.addSuppressed(deleteException);
```

Comment:
1. Yes, I will update the code with more logs about put permissions failing, deleting failing and deleting succeeding.

2. I thought to catch ```Throwable``` when deleting to make sure the final exception thrown to the customer is about putting the permissions that failed, not the deletion. The deletion's exception is just added to not get lost. I consider that only the failed ```PutPermission``` operation  is actionable for the customer.

Reply:
> I thought to catch Throwable when deleting to make sure the final exception thrown to the customer is about putting the permissions that failed, not the deletion. The deletion's exception is just added to not get lost. I consider that only the failed PutPermission operation is actionable for the customer.

It's a reasonable objective, but catching `Throwable` is dangerous since that can include things just as `OutOfMemoryError` and other bits. So we should try to catch a more specific exception.

Reply:
Okay, I will update to catch ```CodeGuruProfilerException``` for deletion.